dynamic_programming/levenshtein_distance.nim

```diff
@@ -0,0 +1,78 @@
+## Levenshtein distance
+
+import tables
+
+type Key = (Natural, Natural)
+
+
+func toKey(indA, indB: int): Key =
+  return (indA.Natural, indB.Natural)
+
+
+func initSubsolutions(lenA, lenB: Natural): Table[Key, Natural] =
+  for indA in 0.Natural..lenA:
+    result[toKey(indA, 0)] = indA
+
+  for indB in 1.Natural..lenB:
+    result[toKey(0, indB)] = indB
+
+
+func computeLevenshteinDistanceMatrix(a, b: string): Table[Key, Natural] =
+  result = initSubsolutions(a.len, b.len)
+
+  for indA in 0..<a.len:
+    for indB in 0..<b.len:
+      let
+        substitutionCost = (if a[indA] == b[indB]: 0 else: 1)
+        distanceIfDeleted = result[toKey(indA, indB + 1)] + 1
+        distanceIfInserted = result[toKey(indA + 1, indB)] + 1
+        distanceIfSubstituted = result[toKey(indA, indB)] + substitutionCost
+      result[toKey(indA + 1, indB + 1)] = min(
+          [distanceIfDeleted,
+           distanceIfInserted,
+           distanceIfSubstituted])
+
+
+func levenshteinDistance*(a, b: string): Natural =
+  ## Returns the
+  ## [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance)
+  ## between the input strings.
+  ## Note that this implementation has O(a.len*b.len) time complexity and
+  ## requires O(a.len*b.len) additional memory
+  runnableExamples:
+    doAssert levenshteinDistance("abc", "aXcY") == 2
+  return computeLevenshteinDistanceMatrix(a, b)[toKey(a.len, b.len)]
+
+
+when isMainModule:
+  import std/unittest
+
+  suite "levenshteinDistance":
+
+    type
+      TestCase = object
+        name, a*, b*: string
+        distance*: Natural
+
+    func getTC(name, a, b: string, distance: int): TestCase =
+      assert distance >= 0
+      return TestCase(name: name, a: a, b: b, distance: distance)
+
+    const testCases = @[
+      getTC("two empty strings", "", "", 0),
+      getTC("empty string vs. one char", "", "a", 1),
+      getTC("two same strings", "b", "b", 0),
+      getTC("one extra character", "a", "ab", 1),
+      getTC("wiki-example 1", "saturday", "sunday", 3),
+      getTC("wiki-example 2", "kitten", "sitting", 3),
+      getTC("4 edits", "abcdefghif", "abCdeFghIfX", 4),
+      getTC("2 edits", "abc", "aXcY", 2),
+      ]
+
+    test "returns expected result":
+      for tc in testCases:
+        check tc.distance == levenshteinDistance(tc.a, tc.b)
+
+    test "is symmetric":
+      for tc in testCases:
+        check tc.distance == levenshteinDistance(tc.b, tc.a)
```

Comment:
I'm sorry I'm still pressing on with this. It's great test cases now have names. The issue with this code is that the user will never see them, as the type is not even participating in the assertion inside `check`s. Moreover, I think you have the structure flipped, instead of having two `test` blocks with loops, it's better to have a single loop with tests, perhaps marked with the `checkpoint("Checkpoint name")` for both possible argument orders. 

> It is a pity that check does not support the syntax like check, condition, message to display then fails

[`unitest.test`](https://nim-lang.org/docs/unittest.html#test.t%2C%2C) is exactly for defining a single named test case with a check[s] inside.

Don't know what you didn't like in my [previous suggestion](https://github.com/TheAlgorithms/Nim/pull/46#discussion_r1242381508). Here's another one, I tried to stay even closer to your code.

```suggestion
  import std/[unittest, sequtils]

  suite "levenshteinDistance":
    type TestCase = object
      name, a, b: string
      distance: Natural

    func toTc(t: tuple[name, a, b: string, distance: int]): TestCase =
      assert t.distance >= 0
      TestCase(name: t.name, a: t.a, b: t.b, distance: Natural(t.distance))

    const testCases = [
      ("two empty strings", "", "", 0),
      ("empty string vs. one char", "", "a", 1),
      ("two same strings", "b", "b", 0),
      ("one extra character", "a", "ab", 1),
      ("wiki-example 1", "saturday", "sunday", 3),
      ("wiki-example 2", "kitten", "sitting", 3),
      ("4 edits", "abcdefghif", "abCdeFghIfX", 4),
      ("2 edits", "abc", "aXcY", 2),
      ].map(toTc)

    for tc in testCases:
      test tc.name:
        checkpoint("A, B")
        check levenshteinDistance(tc.a, tc.b) == tc.distance
        checkpoint("B, B")
        check levenshteinDistance(tc.b, tc.a) == tc.distance
```

Some changes:

0. TestCase name is now properly displayed for each case.
1. TestCase is internal and ad-hoc for this module. There's no need for exports.
2. Don't use returns if not modifyinng the control flow. Please, use [either an expression](https://status-im.github.io/nim-style-guide/language.result.html) or a [`result`](https://nim-lang.org/docs/nep1.html#introduction-coding-conventions) variable.
3. `getTc` renamed to more idiomatic `toTc`: there's no extraction (like from an Option) nor any calculation happening, just a conversion.
4. In my previous suggestion each test case was a named tuple, but here I left them positional and minimal.
  I would go even further and removed the type and the conversion altogether. Then you just replace the `map` with: `].mapIt((name: it[0], a: it[1], b: it[2], distance: Natural(it[3])))`
    Or, if you need the assertion:
    ```nim
    ].mapIt:
      assert it[3]>=0
      (name: it[0], a: it[1], b: it[2], distance: Natural(it[3]))
    ```

Reply:
cf. afc89dd0aae0fa8fc663f146df01301c9d25efe1 (I missed the `test tc.name` part previously).